R/format_data.R

```diff
@@ -2414,6 +2414,331 @@ fmt_datetime <- function(data,
+#' nothing is provided for `rows` then entire columns are selected). Conditional
+#' formatting is possible by providing a conditional expression to the `rows`
+#' argument. See the *Arguments* section for more information on this.
+#'
+#' @inheritParams fmt_number
+#' @param input_units If one or more selected columns contains numeric values,
+#'   the keyword provided for `input_units` will determine how those values are
+#'   interpreted in terms of a single duration type. If numeric columns are
+#'   included in `columns` then one of the following keywords, representative of
+#'   time units, must be provided: seconds (`"secs"`), `"mins"` (minutes),
+#'   `"hours"`, `"days"`, or `"weeks"`.
+#' @param output_units The output values can be handled by one of two keywords
+#'   (`"auto"`, the default, or `"all"`) or a vector of the following time
+#'   parts: `"weeks"`, `"days"`, `"hours"`, `"mins"`, or `"secs"`.
+#' @param duration_style A choice of three formatting styles for the output
+#'   duration values. With `"narrow"` (the default style), durations will be
+#'   formatted with single letter time-part units (e.g., 1.35 days will be
+#'   styled as `"1d 8h 24m`). With `"wide"`, this example value will be expanded
+#'   to `"1 day 8 hours 24 minutes"` after formatting. The `"colon-sep"` style
+#'   of formatting will separate all values with a colon and will not provide
+#'   any time-part units, so the output of 1.35 days will be `"1:8:24"`.
+#' @param drop_zero_parts An option to drop leading and internal values for
+#'   time parts if they are zero. By default this is `TRUE` and duration values
+#'   that might otherwise be formatted as `"0w 1d 0h 4m 19s"` with
+#'   `drop_zero_parts = FALSE` are instead displayed as `"1d 4m 19s"`.
+#'
+#' @return An object of class `gt_tbl`.
+#'
+#' @family Format Data
+#' @section Function ID:
+#' 3-11
+#'
+#' @import rlang
+#' @export
+fmt_duration <- function(
+    data,
+    columns,
+    rows = everything(),
+    input_units = NULL,
+    output_units = "auto",
+    duration_style = c("narrow", "wide", "colon-sep"),
+    drop_zero_parts = TRUE,
+    pattern = "{x}"
+) {
+
+  duration_style <- match.arg(duration_style)
+
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
+  # Stop function if any columns have data that is incompatible
+  # with this formatter
+  if (
+    !column_classes_are_valid(
+      data = data,
+      columns = {{ columns }},
+      valid_classes = c("numeric", "difftime")
+    )
+  ) {
+    stop(
+      "The `fmt_duration()` function can only be used on `columns` of certain types:\n",
+      "* Allowed types are `numeric` and `difftime`",
+      call. = FALSE
+    )
+  }
+
+  # Stop function if any columns have numeric data and `input_units` is NULL
+  if (
+    column_classes_are_valid(
+      data = data,
+      columns = {{ columns }},
+      valid_classes = "numeric"
+    ) &&
```

Comment:
`column_classes_are_valid` currently has `all()` semantics and you need `any()`.